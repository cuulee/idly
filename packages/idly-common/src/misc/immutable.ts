// import { Map as ImMap } from 'immutable';

// export ImMap;
// // export const ImList = Immutable.List;
// // export const ImSet = Immutable.Set;
