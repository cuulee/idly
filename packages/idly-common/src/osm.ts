export type Entity = Node | Way | Relation;
export type Entities = Set<Node | Way | Relation>;
export type EntityId = string;
export type NodeId = string;
export type WayId = string;

export type ParentWays = Map<NodeId, Set<WayId>>;
// Table used to map id -> entity
export type EntityTable = Map<EntityId, Entity>;

export type Tags = Map<string, string>;

export type FeaturePropsTable = Map<EntityId, FeatureProps>;
export type FeatureProps = { [key: string]: string };

export type NodeGeometry =
  | OsmGeometry.POINT
  | OsmGeometry.VERTEX
  | OsmGeometry.VERTEX_SHARED;

export enum OsmGeometry {
  POINT = 'point',
  VERTEX = 'vertex',
  VERTEX_SHARED = 'vertex_shared', // the one shared among multiple ways
  AREA = 'area',
  LINE = 'line',
  RELATION = 'relation'
}

export enum EntityType {
  NODE = 'node',
  WAY = 'way',
  RELATION = 'relation'
}

export interface Node {
  readonly id: EntityId;
  readonly tags: Tags;
  readonly type: EntityType.NODE;
  readonly loc: LngLat;
  readonly attributes: Attributes;
}

export interface Way {
  readonly id: EntityId;
  readonly type: EntityType.WAY;
  readonly tags: Tags;
  readonly attributes: Attributes;
  readonly nodes: EntityId[];
}

export interface Relation {
  readonly id: EntityId;
  readonly type: EntityType.RELATION;
  readonly tags: Tags;
  readonly members: RelationMember[];
  readonly attributes: Attributes;
}

export interface Attributes {
  readonly visible?: boolean;
  readonly version?: number;
  readonly timestamp?: string;
  readonly changeset?: string;
  readonly uid?: string;
  readonly user?: string;
}

export interface RelationMember {
  readonly id: string;
  readonly type: string;
  readonly role: string;
}

export interface LngLat {
  readonly lat: number;
  readonly lon: number;
}
