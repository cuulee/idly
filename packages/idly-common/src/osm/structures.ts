export type Entity = Node | Way | Relation;
export type Entities = Set<Node | Way | Relation>;
export type EntityId = string;
export type NodeId = string;
export type WayId = string;

export interface Tags {
  readonly [key: string]: string;
}

export type FeaturePropsTable = Map<EntityId, FeatureProps>;
export interface FeatureProps {
  [key: string]: string;
}

export type NodeGeometry = OsmGeometry.POINT | OsmGeometry.VERTEX;

export enum OsmGeometry {
  POINT = 'point',
  VERTEX = 'vertex',
  AREA = 'area',
  LINE = 'line',
  RELATION = 'relation',
}

export enum EntityType {
  NODE = 'node',
  WAY = 'way',
  RELATION = 'relation',
}

export interface Node {
  readonly id: EntityId;
  readonly tags: Tags;
  readonly type: EntityType.NODE;
  readonly loc: LngLat;
  readonly attributes: Attributes;
}

export interface Way {
  readonly id: EntityId;
  readonly type: EntityType.WAY;
  readonly tags: Tags;
  readonly attributes: Attributes;
  readonly nodes: ReadonlyArray<EntityId>;
}

export interface Relation {
  readonly id: EntityId;
  readonly type: EntityType.RELATION;
  readonly tags: Tags;
  readonly members: ReadonlyArray<RelationMember>;
  readonly attributes: Attributes;
}

export interface Attributes {
  readonly visible?: boolean;
  readonly version?: string;
  readonly timestamp?: string;
  readonly changeset?: string;
  readonly uid?: string;
  readonly user?: string;
}

// id & ref are the same thing
// id exists for backward compatibility with iD
// we are using ref
export interface RelationMember {
  readonly id: string;
  readonly ref: string;
  readonly type?: string;
  readonly role?: string;
}

export interface LngLat {
  readonly lat: number;
  readonly lon: number;
}
