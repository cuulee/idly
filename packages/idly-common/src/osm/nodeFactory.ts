import { genLngLat } from '../osm/genLngLat';
import { attributesGen } from '../osm/attributesGen';
import {
  Attributes,
  EntityId,
  EntityType,
  LngLat,
  Node,
  Tags
} from '../osm/structures';

export function nodeFactory({
  id,
  tags = new Map(),
  loc = genLngLat([0, 0]),
  attributes = attributesGen({})
}: {
  id: EntityId;
  tags?: Tags;
  loc?: LngLat;
  attributes?: Attributes;
}): Node {
  return {
    id,
    tags,
    type: EntityType.NODE,
    loc,
    attributes
  };
}
