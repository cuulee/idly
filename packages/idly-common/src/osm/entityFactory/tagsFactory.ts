import { sortObjectKeys } from '../../misc/sortObjectKeys';
import { Tags } from '../structures';

// @TOFIX make tags frozen objects.
// and also figure out how to make reserved keys work in object
// or use so other data structure.
const emptyTag = {};
export function tagsFactory(
  rawTags: { [key: string]: string } = emptyTag
): Tags {
  return sortObjectKeys(rawTags);
}
