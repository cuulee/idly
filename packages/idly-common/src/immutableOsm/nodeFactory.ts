import { $attributesGen } from './attributesGen';
import { $genLngLat } from './genLngLat';
import { $tagsFactory } from './tagsFactory';
import { $Node, $__Node } from '../immutableOsm';
import { genLngLat } from '../osm/genLngLat';
import { attributesGen } from '../osm/attributesGen';
import { Attributes, EntityId, EntityType, LngLat, Node, Tags } from '../osm';
import { recordify } from 'typed-immutable-record';

export function nodeFactory(n: Node): $Node {
  var $tags = $tagsFactory(n.tags);
  var $loc = $genLngLat(n.loc);
  var $attr = $attributesGen(n.attributes);
  return recordify<$__Node, $Node>({
    id: n.id,
    tags: $tags,
    type: EntityType.NODE,
    loc: $loc,
    attributes: $attr
  });
}
