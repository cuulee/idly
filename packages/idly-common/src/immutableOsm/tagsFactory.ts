import { Tags } from '../osm/structures';
import { $Tags } from '../immutableOsm/immutableOsm';
import { Map as $Map } from 'immutable';

export function $tagsFactory(t: Tags): $Tags {
  return $Map(t);
}
