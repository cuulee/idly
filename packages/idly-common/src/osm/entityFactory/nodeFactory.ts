import { lngLatFactory } from '../../geo/lngLatFactory';
import { deepFreeze } from '../../misc/deepFreeze';

import {
  Attributes,
  EntityId,
  EntityType,
  LngLat,
  Node,
  Tags,
} from '../structures';
import { attributesFactory } from './attributesFactory';
import { tagsFactory } from './tagsFactory';

export function nodeFactory(
  {
    id,
    tags = tagsFactory(),
    loc = lngLatFactory([0, 0]),
    attributes = attributesFactory(),
  }: {
    id: EntityId;
    tags?: Tags;
    loc?: LngLat;
    attributes?: Attributes;
  },
  freeze = true
): Node {
  return deepFreeze<Node>(
    {
      attributes: attributesFactory(attributes),
      id,
      loc,
      tags: tagsFactory(tags),
      type: EntityType.NODE,
    },
    freeze
  );
}
