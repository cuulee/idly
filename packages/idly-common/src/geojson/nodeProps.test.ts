import { Map as ImMap, Set as ImSet } from 'immutable';

import {
  applyNodeMarkup,
  DEFAULT_NODE_ICON,
  nodePropertiesGen,
} from './nodeProps';

import { nodeFactory } from '../osm/nodeFactory';
import { ParentWays } from '../osm/structures';
import { OsmGeometry } from '../osm/structures';
import { tagsFactory } from '../osm/tagsFactory';

export function dummyParentWaysGen(obj: any): ParentWays {
  return ImMap(obj);
}

const n1 = nodeFactory({ id: 'n-1' });

const dummyParentWays = dummyParentWaysGen({
  n3780767744: ImSet(['w40542208']),
  n4558992269: ImSet(['w40542208']),
  n253179996: ImSet(['w40882200', 'w237684574', 'w173431854', 'w450548831']),
  n1485636774: ImSet(['w40882200', 'w135262258']),
});
describe('node property generator', () => {
  it('should work', () => {
    // @TOFIX change to immutable once finsished rolling out
    expect(
      nodePropertiesGen(n1, dummyParentWays.get(n1.id) || ImSet()),
    ).toMatchSnapshot();
  });

  describe('applyNodeMarkup', () => {
    it('should work', () => {
      const markup = applyNodeMarkup(
        OsmGeometry.POINT,
        tagsFactory({ k: 'k' }),
      );
      expect(markup).toEqual({
        icon: DEFAULT_NODE_ICON,
        name: undefined,
        geometry: OsmGeometry.POINT,
      });
    });
    it('should get the icon', () => {
      const markup = applyNodeMarkup(
        OsmGeometry.POINT,
        tagsFactory({ k: 'k' }),
      );
      expect(markup).toEqual({
        icon: DEFAULT_NODE_ICON,
        name: undefined,
        geometry: OsmGeometry.POINT,
      });
    });
    it('should get the name', () => {
      const markup = applyNodeMarkup(
        OsmGeometry.POINT,
        tagsFactory({ name: 'whatsInTheName' }),
      );
      expect(markup).toEqual({
        icon: DEFAULT_NODE_ICON,
        name: 'whatsInTheName',
        geometry: OsmGeometry.POINT,
      });
    });
  });
});
