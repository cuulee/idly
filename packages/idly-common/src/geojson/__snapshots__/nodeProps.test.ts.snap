// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`node property generator should work 1`] = `
Object {
  "geometry": "point",
  "icon": "circle",
  "name": undefined,
}
`;
